src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4345,32 +4578,101 @@ private uint GetTupleValEscape(ImmutableArray<BoundExpression> elements, uint sc
+        /// as a stored field.
+        /// </summary>
         private uint GetValEscapeOfObjectInitializer(BoundObjectInitializerExpression initExpr, uint scopeOfTheContainingExpression)
         {
             var result = CallingMethodScope;
-            foreach (var expression in initExpr.Initializers)
+            foreach (var expr in initExpr.Initializers)
             {
-                if (expression.Kind == BoundKind.AssignmentOperator)
-                {
-                    var assignment = (BoundAssignmentOperator)expression;
-                    var rightValEscape = assignment.IsRef
-                        ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
-                        : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
+                var exprResult = GetValEscapeOfObjectMemberInitializer(expr, scopeOfTheContainingExpression);
+                result = Math.Max(result, exprResult);
+            }
 
-                    result = Math.Max(result, rightValEscape);
+            return result;
+        }
+
+        private uint GetValEscapeOfObjectMemberInitializer(BoundExpression expr, uint scopeOfTheContainingExpression)
+        {
+            uint result;
+            if (expr.Kind == BoundKind.AssignmentOperator)
+            {
+                var assignment = (BoundAssignmentOperator)expr;
+                var rightEscape = assignment.IsRef
+                    ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
+                    : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
 
-                    var left = (BoundObjectInitializerMember)assignment.Left;
-                    result = Math.Max(result, GetValEscape(left.Arguments, scopeOfTheContainingExpression));
+                uint leftEscape;
+                var left = (BoundObjectInitializerMember)assignment.Left;
+                if (left.MemberSymbol is PropertySymbol { IsIndexer: true } property)
+                {
+                    Debug.Assert(left.AccessorKind != AccessorKind.Unknown);
+                    MethodInfo methodInfo = MethodInfo.Create(property, left.AccessorKind);
+                    leftEscape = getIndexerEscape(methodInfo, left, rightEscape);
                 }
                 else
                 {
-                    result = Math.Max(result, GetValEscape(expression, scopeOfTheContainingExpression));
+                    leftEscape = GetValEscape(left.Arguments, scopeOfTheContainingExpression);
+                    leftEscape = Math.Max(leftEscape, rightEscape);
                 }
+
+                result = leftEscape;
+            }
+            else
+            {
+                result = GetValEscape(expr, scopeOfTheContainingExpression);
             }
 
             return result;
+
+            uint getIndexerEscape(
+                in MethodInfo methodInfo,
+                BoundObjectInitializerMember expr,
+                uint rightEscapeScope)
```

Comment:
>rightEscapeScope [](http://example.com/codeflow?start=21&length=16)

Is this argument necessary, or should the caller use `Math.Max(getIndexerEscape(...), rightEscapeScope)`?

Reply:
Actually that would treat the escape scope as `rightEscapeScope` when `method.IsEffectivelyReadOnly`, which seems incorrect.